cmd/cspc-operator/app/cleanup.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2020 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	bdc "github.com/openebs/maya/pkg/blockdeviceclaim/v1alpha1"
+	apiscspc "github.com/openebs/maya/pkg/cstor/poolcluster/v1alpha1"
+	cspi "github.com/openebs/maya/pkg/cstor/poolinstance/v1alpha3"
+	util "github.com/openebs/maya/pkg/util"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+// cleanupCSPIResources removes the CSPI resources when a CSPI is
+// deleted or downscaled
+func cleanupCSPIResources(cspcObj *apis.CStorPoolCluster) error {
+	cspiList, err := cspi.NewKubeClient().WithNamespace(cspcObj.Namespace).List(
+		metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + cspcObj.Name,
+		},
+	)
+	if err != nil {
+		return errors.Errorf("failed to list cspi for cspc %s to perform cleanup: %s", cspcObj.Name, err.Error())
+	}
+	opts := []cspiCleanupOptions{cleanupBDC}
+	for _, cspiItem := range cspiList.Items {
+		cspiItem := cspiItem // pin it
+		cspiObj := &cspiItem
+		// cleanup to be performed only if DeletionTimestamp is non zero and if
+		// PoolProtectionFinalizer is not removed wait for the next reconcile attempt
+		if canPerformCSPICleanup(cspiObj) {
```

Comment:
please bring back the check for deleteTimestamp, otherwise, this function returns error for most of the times and reconciliation may not work.

Reply:
The check is still present as a new predicate in the canPerformCSPICleanup function.